apf/apfH1Shapes.cc

```diff
@@ -0,0 +1,769 @@
+      double w = cp[i] + cp[j] + cp[p-i-j];
+      nodes[o][0] = cp[p-i-j]/w;  nodes[o][1] = cp[i]/w;  nodes[o][2] = cp[j]/w;
+      o++;
+    }
+
+  // (0,2,3)
+  for (int j = 1; j < p; j++)
+    for (int i = 1; i + j < p; i++) {
+      double w = cp[i] + cp[j] + cp[p-i-j];
+      nodes[o][0] = cp[0];  nodes[o][1] = cp[i]/w;  nodes[o][2] = cp[j]/w;
+      o++;
+    }
+
+  // Region
+  for (int k = 1; k < p; k++) {
+    for (int j = 1; j + k < p; j++) {
+      for (int i = 1; i + j + k < p; i++) {
+        double w = cp[i] + cp[j] + cp[k] + cp[p-i-j-k];
+        nodes[o][0] = cp[i]/w;  nodes[o][1] = cp[j]/w;  nodes[o][2] = cp[k]/w;
+        o++;
+      }
+    }
+  }
+
+  // Populate T
+  mth::Matrix<double> T(non,non);
+  for (int m = 0; m < non; m++)
+  {
+    o = 0;
+    double x = nodes[m][0]; double y = nodes[m][1]; double z = nodes[m][2];
+
+    getChebyshevT(p, x, &shape_x[0]);
+    getChebyshevT(p, y, &shape_y[0]);
+    getChebyshevT(p, z, &shape_z[0]);
+    getChebyshevT(p, 1. - x - y - z, &shape_l[0]);
+
+    for (int k = 0; k <= p; k++)
+      for (int j = 0; j + k <= p; j++)
+        for (int i = 0; i + j + k <= p; i++)
+          T(o++, m) = shape_x[i]*shape_y[j]*shape_z[k]*shape_l[p-i-j-k];
+	}
+  mth::decomposeQR(T, Q, R);
+}
+
+static void getTi(
+    int P,
+    int type,
+    mth::Matrix<double>& Q,
+    mth::Matrix<double>& R)
+{
+
+  bool cond = (type == apf::Mesh::TRIANGLE || type == apf::Mesh::TET);
+  PCU_ALWAYS_ASSERT_VERBOSE(cond,
+      "type should be either apf::Mesh::TRIANGLE or apf::Mesh::TET!");
+
+  static apf::NewArray<double> transformQ[apf::Mesh::TYPES][MAX_ORDER+1];
+  static apf::NewArray<double> transformR[apf::Mesh::TYPES][MAX_ORDER+1];
+  int n = type == apf::Mesh::TRIANGLE ? countTriNodes(P) : countTetNodes(P);
+
+  // get the transform matrices if the are not already computed
+  if (!transformQ[type][P].allocated()) {
+    mth::Matrix<double> LQ(n,n);
+    mth::Matrix<double> LR(n,n);
+    type == apf::Mesh::TRIANGLE ?
+    	  computeTriangleTi(P, LQ, LR) : computeTetTi(P, LQ, LR);
+
+    transformQ[type][P].allocate(n*n);
+    transformR[type][P].allocate(n*n);
+
+    for (int i = 0; i < n; i++) {
+      for (int j = 0; j < n; j++) {
+	      transformQ[type][P][i*n+j] = LQ(i,j);
+      	transformR[type][P][i*n+j] = LR(i,j);
+      }
+    }
+  }
+
+  // set Q and R using transformQ and transformR
+  for (int i = 0; i < n; i++) {
+    for (int j = 0; j < n; j++) {
+      Q(i,j) = transformQ[type][P][i*n+j];
+      R(i,j) = transformR[type][P][i*n+j];
+    }
+  }
+}
+
+// internal nodes only
+static apf::Vector3 getH1NodeXi(int type, int P, int node)
+{
+  if (type == apf::Mesh::VERTEX)
+    return apf::Vector3(0., 0., 0.);
+
+  apf::NewArray<double> cp;
+  getClosedPoints(P, cp);
+
+  if (type == apf::Mesh::EDGE) {
+    PCU_ALWAYS_ASSERT(node >= 0 && node < countInternalEdgeNodes(P));
+    int c = 0;
+    for (int i = 1; i < P; i++)
+      if (node == c)
+      	return apf::Vector3(2*cp[i]-1, 0., 0.);
```

Comment:
tabs?  github does not always render the text white space correctly

Reply:
Fixed this.